spec/errors.md

```diff
@@ -16,10 +16,10 @@ and MUST be emitted as soon as possible.
 The other error categories are only emitted during formatting,
 but it might be possible to detect them with validation tools.
 
-During selection, an _expression_ handler MUST only emit _Resolution Errors_ and _Selection Errors_.
-During formatting, an _expression_ handler MUST only emit _Resolution Errors_ and _Formatting Errors_.
+During selection and formatting,
+_expression_ handlers MUST only emit _Message Function Errors_.
```

Comment:
Maybe murder this normative text?

The MUST isn't testable. It would be okay to have a normative statement about what errors can be emitted in a specific phase of processing. But the boundaries of 'selection' and 'formatting' are vague _in this context_.

Perhaps try turning this upside down:

```suggestion
_Message Function Errors_ MUST only be emitted during selection and formatting.
```

Reply:
The intent is to require all emitted errors to be Message Function Errors, rather than any random errors. The phrasing is perhaps a bit awkward as this used to have two separate pharses talking about selection and formatting. Would this work?

```suggestion
An _expression_ handler MUST only emit _Message Function Errors_.
```

Reply:
I think my comment still applies? If our set of errors is more-or-less comprehensive and our instructions are clear about when to emit specific errors, we don't need to have normative text forbidding implementers from choosing one or another error type or about when they emit those errors. It's impossible to test anyway.

We'd be better off with non-normative text that signals our intention.

We talk about syntax/data model errors specifically and about message function errors specifically in this section, but only mention resolution errors in passing.

I think I'd rewrite this section (including the paragraph above that I can't `suggestion` into submission...) along the lines of:

```suggestion
_Syntax Errors_ and _Data Model Errors_ can always be detected during message parsing or processing.
(Other categories of error can sometimes be detected using validation tools.)

_Resolution Errors_ can be detected during _variable_, _function_, or _markup_ resolution.

_Message Function Errors_ apply to _selection_ or _formatting_
and include errors detected during the evaluation of an
_expression_ or during the execution of a _function_.
```

Reply:
Note the preceding paragraph, which already talks of catching errors with validation tools. So repeating that here seems rather odd.

Reply:
The "suggestion" block is intended to replace the preceding paragraph also. So it's not a repeat.

Reply:
So you're proposing that we also drop the requirement to emit syntax & data model errors as early as possible?

Thus far, this PR is only about changing the names & categorisation of errors, which is already a lot. Separately we also have #704 attempting to change a related aspect of error handling.

Could we leave error handling changes for separate consideration? This seems like a topic we'll need to discuss more, and I would much rather not get the rest of this PR blocked on that.

If that's acceptable, please feel free to resolve this thread by accepting my last suggestion above, or by closing it and keeping the current language.

Reply:
Yeah, I'm proposing that we spend more time on error handling, noting that there's a problem here. I'll create an issue to track that proposal so that we can merge this.